# Summarize Search Results

The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:

---
- **<a href="../../QuickStart/usage/#q">Q.()</a>** builds a query that can be used by `run()` or `count()`
- **<a href="../../QuickStart/usage/#qrun">Q.run()</a>** returns data for the specified search 
- **<a href="../../QuickStart/usage/#qcount">Q.count()</a>** returns summary information (counts) data that fit the specified search
- **<a href="../../QuickStart/usage/#columns">columns()</a>** returns entity field names
- **<a href="../../QuickStart/usage/#unique_terms">unique_terms()</a>** returns entity field contents

---
                                                                    
Before we do any work, we needs to import these functions cdapython.
We're also telling cdapython to report it's version so we can be sure we're using the one we mean to:

In [3]:
from cdapython import Q, columns, unique_terms, query
print(Q.get_version())
Q.set_host_url('http://localhost:8080')
Q.set_default_project_dataset("gdc-bq-sample.dev")

2022.8.29

<div class="cdanote" style="background-color:#b3e5d5;color:black;padding:20px;">
    
CDA data comes from three sources:
<ul>
<li><b>The <a href="https://proteomic.datacommons.cancer.gov/pdc/"> Proteomic Data Commons</a> (PDC)</b></li>
<li><b>The <a href="https://gdc.cancer.gov/">Genomic Data Commons</a> (GDC)</b></li>
<li><b>The <a href="https://datacommons.cancer.gov/repository/imaging-data-commons">Imaging Data Commons</a> (IDC)</b></li>
</ul> 
    
The CDA makes this data searchable in four main endpoints:

<ul>
<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>
<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>
<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>
<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>
</ul>
    
And two endpoints that offer deeper information about data in the researchsubject endpoint:
<ul>
<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>
<li><b>treatment:</b> Represent medication administration or other treatment types.</li>
</ul>
Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.
</div>


If you are looking to build a cohort of distinct individuals who meet some criteria, search by `subject`. If you want to build a cohort, but are particularly interested in studies rather than the participates per se, search by `researchsubject`. If you are looking for biosamples that can be ordered or a specific format of information (for e.g. histological slides) start with `specimen`. If you are primarily looking for files you can reuse for your own analysis, start with `file`.

In CDA search, these concepts can also be chained together, so you can look specifically for specimen subjects, or researchsubject diagnoses. In the four 'main' tables, all of the rows will have one or more files associated with them that can be directly found by chaining, as in specimen files. Diagnosis and treatment do not have files directly associated with them and so can only be used to find files in conjunction with the other searches.

In all cases, any search can use any metadata field, the only difference between search types is what type of data you return by default. 



## Getting simple summary data

Let's try a broad search of the CDA to see what information exists about cancers that were first diagnosed in the brain. To run this simple search, we would first construct a query in `Q` and save it to a variable `myquery`. This is the same query we ran in the <a href="../BasicSearch">Basic Search</a> notebook:

In [4]:



myquery = Q('primary_diagnosis_site = "brain"')


<div class="cdawarn" style="background-color:#f9cfbf;color:black;padding:20px;">
<h3>Where did those terms come from?</h3>
    
If you aren't sure how we knew what terms to put in our search, please refer back to the <a href="../SearchTerms">What search terms are available?</a> notebook. 
</div>

### Overall summary

You can get a quick summary of how many unique specimens, treatments, diagnoses, researchsubjects and subjects meet your search criteria by chaining a `count` command into the basic `run` call. 

In [5]:
myquery.count.run()

Getting results from database

Waiting for results

Total execution time: 0
                        min 15.935 sec 15935 ms

specimen_count : 39201

treatment_count : 2386

diagnosis_count : 1756

mutation_count : 904

researchsubject_count : 3716

subject_count : 2384

These numbers are how many total rows of data will come back when querying the various endpoints.



### subject summary

We can also add `count`to the other run calls we did in the <a href="../BasicSearch">Basic Search</a> notebook to get more detailed summaries:

In [8]:
myquery.subject.count.run(show_sql=True)

Getting results from database

Waiting for results

Total execution time: 0
                        min 33.746 sec 33746 ms

total : 2384

files : 4099497

subject_identifier_system 
 count 
 
 
 
 
 IDC 
 1955 
 
 
 PDC 
 309 
 
 
 GDC 
 1454 
 
 
 

 
 
 
 sex 
 count 
 
 
 
 
 None 
 748 
 
 
 female 
 653 
 
 
 male 
 980 
 
 
 not reported 
 3 
 
 
 

 
 
 
 race 
 count 
 
 
 
 
 None 
 748 
 
 
 white 
 1311 
 
 
 black or african american 
 96 
 
 
 not reported 
 136 
 
 
 asian 
 33 
 
 
 not allowed to collect 
 25 
 
 
 american indian or alaska native 
 4 
 
 
 Unknown 
 21 
 
 
 other 
 9 
 
 
 native hawaiian or other pacific islander 
 1 
 
 
 

 
 
 
 ethnicity 
 count 
 
 
 
 
 None 
 748 
 
 
 not hispanic or latino 
 1285 
 
 
 not reported 
 219 
 
 
 hispanic or latino 
 85 
 
 
 Unknown 
 22 
 
 
 not allowed to collect 
 25 
 
 
 

 
 
 
 cause_of_death 
 count 
 
 
 
 
 None 
 2098 
 
 
 Not Reported 
 200 
 
 
 Cancer Related 
 63 
 
 
 Not Cancer Related 
 9 
 
 
 Infection 
 3 
 
 
 Unknown 
 9 
 
 
 Surgical Complications 
 2

with flattened_result as (SELECT results.* EXCEPT(rn) FROM (SELECT ROW_NUMBER() OVER (PARTITION BY Subject.id,     
_subject_identifier.system, _subject_associated_project, _subject_Files) as rn, Subject.id AS subject_id,          
_subject_identifier.system AS subject_identifier_system, _subject_identifier.value AS subject_identifier_value,    
Subject.species AS species, Subject.sex AS sex, Subject.race AS race, Subject.ethnicity AS ethnicity,              
Subject.days_to_birth AS days_to_birth, _subject_associated_project, Subject.vital_status AS vital_status,         
Subject.days_to_death AS days_to_death, Subject.cause_of_death AS cause_of_death, _subject_Files FROM              
gdc-bq-sample.dev.all_Subjects_v3_0_final AS Subject LEFT JOIN UNNEST(Subject.ResearchSubject) AS _ResearchSubject 
LEFT JOIN UNNEST(Subject.identifier) AS _subject_identifier LEFT JOIN UNNEST(Subject.subject_associated_project) AS
_subject_associated_project LEFT JOIN UNNEST(Subject.Files) AS _subject_Files WHERE                                
(IFNULL(UPPER(_ResearchSubject.primary_diagnosis_site), '') = UPPER('brain'))) as results WHERE rn = 1) select     
(SELECT COUNT(DISTINCT subject_id) from flattened_result) as total, (SELECT COUNT(DISTINCT _subject_Files) from    
flattened_result) as files, (select ARRAY(select as STRUCT subject_identifier_system, count(distinct subject_id) as
count from flattened_result group by subject_identifier_system)) as subject_identifier_system, (select ARRAY(select
as STRUCT sex, count(distinct subject_id) as count from flattened_result group[38;2;248;248;242;48;2;39;40;3

Since we save the output as a variable, we need to look at the variable to see the results:

In [5]:
subjectresults

By default, the results are displayed as a table for easy previewing of the data. Since we queried the `subject` endpoint, our default results tell us `subject` level information, that is, information about unique individuals: their sex, race, age, species, etc. Using counts gives us back a nice pivot table type summary of the countable fields for Subjects. Note that above the table it also tells you the total subject count, as well as how many files are associated with those subjects.


---

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>Subject Field Definitions</h3>

<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>
    
    
    
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-7zrl{text-align:left;vertical-align:bottom}
.tg .tg-0lax{text-align:left;vertical-align:top}
</style>
<table class="tg">
<tbody>
  <tr>
    <td class="tg-0lax"> id (`total`)</td>
    <td class="tg-0lax"> The overall number of subjects returned.</td>
  </tr>
  <tr>
    <td class="tg-0lax"> files</td>
    <td class="tg-0lax"> The number of files that match this search.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">identifier.value(`system`)</td>
    <td class="tg-0lax"> The identifier for the data provider.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">species</td>
    <td class="tg-0lax"> The taxonomic group (e.g. species) of the subject.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">sex</td>
    <td class="tg-0lax">  The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics. </td>
  </tr>
  <tr>
    <td class="tg-7zrl">race</td>
    <td class="tg-0lax"> An arbitrary classification of a taxonomic group that is a division of a species.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">ethnicity</td>
    <td class="tg-0lax"> An individual's self-described social and cultural grouping.</td>
  </tr>
  <tr>
    <td class="tg-0lax"> cause_of_death</td>
    <td class="tg-0lax"> The cause of death, if known</td>
  </tr>
</tbody>
</table>

</div>
    
---

This gives you a quick way to assess whether the full search results will have the data fields you require. But if you want to get the underlying data for your own downstream applications, you can also get the raw numbers by calling the zeroth value of the variable:

In [6]:
subjectresults[0]

TypeError: 'NoneType' object is not subscriptable

### researchsubject

If we're interested in what researchsubjects meet our criteria, we can also run our query against the researchsubject endpoint. Lets run it without saving to a variable this time to make it a bit quicker:

In [8]:
myquery.researchsubject.count.run()

Getting results from database

Waiting for results

Http Status: 500
                Error Message: Unrecognized name: _researchsubject_identifier at [1:121]

Total execution time: 0
                        min 1.08 sec 1080 ms




---

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>ResearchSubject Field Definitions</h3>

<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>
    
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-7zrl{text-align:left;vertical-align:bottom}
.tg .tg-0lax{text-align:left;vertical-align:top}
</style>
<table class="tg">
<tbody>
  <tr>
    <td class="tg-0lax"> id (`total`)</td>
    <td class="tg-0lax"> The overall number of researchsubjects returned.</td>
  </tr>
  <tr>
    <td class="tg-0lax"> files</td>
    <td class="tg-0lax"> The number of files that match this search.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">identifier.value(`system`)</td>
    <td class="tg-0lax"> The identifier for the data provider.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">primary_diagnosis_condition</td>
    <td class="tg-0lax"> The text term used to describe the type of malignant disease.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">primary_diagnosis_site</td>
    <td class="tg-0lax"> The text term used to describe the primary site of disease.</td>
  </tr>
</tbody>
</table>
</div>
    
---

### diagnosis

The diagnosis endpoint is an extension of the researchsubject endpoint, and returns information about researchsubjects that have a diagnosis that meets our search criteria. :

In [9]:
myquery.diagnosis.count.run()

Getting results from database

Waiting for results

Http Status: 500
                Error Message: Unrecognized name: _diagnosis_identifier at [1:115]

Total execution time: 0
                        min 0.942 sec 942 ms

---

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>Diagnosis Field Definitions</h3>

<i>A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</i>

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-za14{border-color:inherit;text-align:left;vertical-align:bottom}
.tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}
.tg .tg-7zrl{text-align:left;vertical-align:bottom}
.tg .tg-0lax{text-align:left;vertical-align:top}
</style>
<table class="tg">
<tbody>
  <tr>
    <td class="tg-0lax"> id (`total`)</td>
    <td class="tg-0lax"> The overall number of diagnoses returned.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">identifier.value(`system`)</td>
    <td class="tg-0lax"> The identifier for the data provider.</td>
  </tr>
  <tr>
    <td class="tg-za14">primary_diagnosis</td>
    <td class="tg-0pky"> The diagnosis instance that qualified a subject for inclusion on a ResearchProject.</td>
  </tr>
  <tr>
    <td class="tg-za14">stage</td>
    <td class="tg-0pky"> The extent of a cancer in the body.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">grade</td>
    <td class="tg-0lax"> The degree of abnormality of cancer cells.</td>
  </tr>
</tbody>
</table>


</div>
    
---


### treatment

The treatment endpoint is an extension of diagnosis and returns information about treatments undertaken on research subjects that have a given diagnosis that meets our search criteria:

In [10]:
myquery.treatment.count.run()

Getting results from database

Waiting for results

Http Status: 500
                Error Message: Unrecognized name: _treatment_identifier at [1:115]

Total execution time: 0
                        min 0.732 sec 732 ms


---

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>Treatment Field Definitions</h3>

<i><i> Medication administration or other treatment types. A single research subject may have multiple treatments for a single diagnosis, and/or different diagnoses, and different treatments, across different studies</i></i>
    
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-za14{border-color:inherit;text-align:left;vertical-align:bottom}
.tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}
.tg .tg-7zrl{text-align:left;vertical-align:bottom}
.tg .tg-0lax{text-align:left;vertical-align:top}
</style>
<table class="tg">
<tbody>
  <tr>
    <td class="tg-0lax"> id (`total`)</td>
    <td class="tg-0lax"> The overall number of treatments returned.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">identifier.value(`system`)</td>
    <td class="tg-0lax"> The identifier for the data provider.</td>
  </tr>
  <tr>
    <td class="tg-za14">treatment_type</td>
    <td class="tg-0pky"> The treatment type including medication/therapeutics or other procedures.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">treatment_effect</td>
    <td class="tg-0lax">The effect of a treatment on the diagnosis or tumor. </td>
  </tr>
</tbody>
</table>
 

</div>
    
---




### specimens

We can use this same query to see what specimens are available for brain tissue at the CDA:

In [11]:
myquery.specimen.count.run()

Getting results from database

Waiting for results

Http Status: 500
                Error Message: Unrecognized name: _specimen_identifier at [1:114]

Total execution time: 0
                        min 0.766 sec 766 ms

Nearly 40,000 specimens with over 50,000 files meet our search criteria! We would typically expect this number to be much larger than our number of subjects or research_subjects. First because studies will often take more than one sample per subject, and second because any given specimen might be aliquoted out to be used in multiple tests. 

<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>Specimen Field Definitions</h3>

<i><i>Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</i>
 A given specimen will have only a single subject ID and a single research subject ID</i>
    
    
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-za14{border-color:inherit;text-align:left;vertical-align:bottom}
.tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}
.tg .tg-7zrl{text-align:left;vertical-align:bottom}
.tg .tg-0lax{text-align:left;vertical-align:top}
</style>
    
<table class="tg">
<tbody>
  <tr>
    <td class="tg-0lax"> id (`total`)</td>
    <td class="tg-0lax"> The overall number of specimens returned.</td>
  </tr>
  <tr>
    <td class="tg-0lax"> files</td>
    <td class="tg-0lax"> The number of files that match this search.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">identifier.value(`system`)</td>
    <td class="tg-0lax"> The identifier for the data provider.</td>
  </tr>
  <tr>
    <td class="tg-za14">primary_disease_type</td>
    <td class="tg-0pky"> The text term used to describe the type of malignant disease.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">source_material_type</td>
    <td class="tg-0lax"> The general kind of material from which the specimen was derived.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">specimen_type</td>
    <td class="tg-0lax"> The high-level type of the specimen, based on its how it has been derived from the original extracted sample. One of: analyte, aliquot, portion, sample, or slide.</td>
  </tr>
</tbody>
</table>
</div>

### file

The file endpoint returns all files that match our query:

In [12]:
myquery.file.count.run()

Getting results from database

Waiting for results

Http Status: 500
                Error Message: None

Total execution time: 0
                        min 0.104 sec 104 ms

There are a huge number of files (4099497) that match our search. Likely we would want to additionally filter the results by file format or data type to get only files we can use. See all the ways you can filter and refine searches  with more search terms in the <a href="../Operators">Operators</a> notebook.


<div class="cdadefine" style="background-color:#add9e5;color:black;padding:20px;">

<h3>File Field Definitions</h3>

<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>

    
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:noto sans, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-7zrl{text-align:left;vertical-align:bottom}
</style>
<table class="tg">
<tbody>
  <tr>
    <td class="tg-0lax"> id (`total`)</td>
    <td class="tg-0lax"> The overall number of files returned.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">identifier.value(`system`)</td>
    <td class="tg-0lax"> The identifier for the data provider.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">data_catagory</td>
    <td class="tg-7zrl">Broad categorization of the contents of the data file.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">data_type</td>
    <td class="tg-7zrl">Specific content type of the data file.</td>
  </tr>
  <tr>
    <td class="tg-7zrl">file_format</td>
    <td class="tg-7zrl">Format of the data files.</td>
  </tr>
</tbody>
</table>

</div>

### mutation

The mutation endpoint returns all mutations that match our query:

In [9]:
myquery.mutation.count.run()

Getting results from database

Waiting for results

Total execution time: 0
                        min 8.545 sec 8545 ms

total : 904

KeyError: 'ncbi_build'

## Files from a single endpoint (endpoint chaining)

If you want all file formats and data types, but only from a specific endpoint, you can also filter the file results by chaining endpoints together. This will return all the files that match our search AND that are specifically from specimens:

In [14]:
myquery.specimen.file.count.run()

Getting results from database

Waiting for results

Http Status: 500
                Error Message: None

Total execution time: 0
                        min 0.117 sec 117 ms

Learn more about chaining endpoints in the [Chaining endpoints]("../AdvancedSearch-Chaining") notebook.